src/ccl_tracers.c

```diff
@@ -58,40 +61,62 @@ static double get_nz_norm(ccl_cosmology *cosmo, ccl_f1d_t *nz_f,
                           double z0, double zf, int *status) {
   double nz_norm = -1, nz_enorm;
 
-  // Get N(z) norm
-  gsl_function F;
-  gsl_integration_workspace *w = NULL;
-  F.function = &nz_integrand;
-  F.params = nz_f;
-
-  w = gsl_integration_workspace_alloc(cosmo->gsl_params.N_ITERATION);
-
-  if (w == NULL) {
-    *status = CCL_ERROR_MEMORY;
-    ccl_cosmology_set_status_message(
-      cosmo, "ccl_tracers.c: get_nz_norm(): out of memory");
+  if(cosmo->gsl_params.NZ_NORM_SPLINE_INTEGRATION) {
+    gsl_interp_accel *accel = NULL;
+    accel = gsl_interp_accel_alloc();
+    if(accel == NULL) {
+      *status = CCL_ERROR_MEMORY;
+      ccl_cosmology_set_status_message(
+        cosmo, "ccl_tracers.c: get_nz_norm(): out of memory");
+    }
+    if(*status == 0) {
+      *status = gsl_spline_eval_integ_e(nz_f->spline, z0, zf, accel, &nz_norm);
+      if(*status) {
+        *status = CCL_ERROR_SPLINE_EV;
+        nz_norm = NAN;
+        ccl_cosmology_set_status_message(
+          cosmo, "ccl_tracers.c: get_nz_norm(): Spline integration failed.");
+      }
+    }
+    gsl_interp_accel_free(accel);
+    return nz_norm;
   }
   else {
-    int gslstatus = gsl_integration_qag(
-      &F, z0, zf, 0,
-      cosmo->gsl_params.INTEGRATION_EPSREL,
-      cosmo->gsl_params.N_ITERATION,
-      cosmo->gsl_params.INTEGRATION_GAUSS_KRONROD_POINTS,
-      w, &nz_norm, &nz_enorm);
-
-    if (gslstatus != GSL_SUCCESS) {
-      ccl_raise_gsl_warning(gslstatus, "ccl_tracers.c: get_nz_norm():");
-      *status = CCL_ERROR_INTEG;
+    // Use GSL integration routine
+    gsl_function F;
+    gsl_integration_workspace *w = NULL;
+    F.function = &nz_integrand;
+    F.params = nz_f;
+
+    w = gsl_integration_workspace_alloc(cosmo->gsl_params.N_ITERATION);
+
+    if (w == NULL) {
+      *status = CCL_ERROR_MEMORY;
       ccl_cosmology_set_status_message(
-        cosmo,
-        "ccl_tracers.c: get_nz_norm(): "
-        "integration error when normalizing N(z)\n");
+        cosmo, "ccl_tracers.c: get_nz_norm(): out of memory");
+    }
+    else {
+      int gslstatus = gsl_integration_qag(
+        &F, z0, zf, 0,
+        cosmo->gsl_params.INTEGRATION_EPSREL,
+        cosmo->gsl_params.N_ITERATION,
+        cosmo->gsl_params.INTEGRATION_GAUSS_KRONROD_POINTS,
+        w, &nz_norm, &nz_enorm);
+
+      if (gslstatus != GSL_SUCCESS) {
+        ccl_raise_gsl_warning(gslstatus, "ccl_tracers.c: get_nz_norm():");
+        *status = CCL_ERROR_INTEG;
+        ccl_cosmology_set_status_message(
+          cosmo,
+          "ccl_tracers.c: get_nz_norm(): "
+          "integration error when normalizing N(z)\n");
+      }
```

Comment:
Why are we raising both a warning and an error here?

Reply:
No reason. Just some left-over I had forgotten to remove earlier.